src/test/java/com/atypon/wayf/integration/publisher/PublisherDeviceIntegrationTest.java

```diff
@@ -0,0 +1,207 @@
+
+        // Get the maximum last active active date
+        Date latestLastActiveDate = ResponseWriter.DATE_FORMAT.parse(DATE_FORMAT.format(new Date()));
+
+        // Test the device history after adding the IDPs
+        String deviceHistoryFromPublisherA = deviceAccessTestUtil.testDeviceHistory(publisherALocalId, publisherAToken, INITIAL_ADD_IDP_DEVICE_HISTORY_RESPONSE_JSON);
+        deviceAccessTestUtil.testLastActiveDateBetween(earliestLastActiveDate, latestLastActiveDate, 3, deviceHistoryFromPublisherA);
+
+        // Relate the device to publisher B
+        String publisherBLocalId = "local-id-publisher-b-" + UUID.randomUUID().toString();
+        String globalIdPublisherB = deviceTestUtil.relateDeviceToPublisher(publisherBLocalId, publisherBToken, globalIdPublisherA, RELATE_NEW_DEVICE_PUBLISHER_A_RESPONSE_JSON);
+
+        assertEquals(globalIdPublisherA, globalIdPublisherB);
+
+        // Get the usage history for publisher B
+        String deviceHistoryFromPublisherB = deviceAccessTestUtil.testDeviceHistory(publisherBLocalId, publisherBToken, INITIAL_ADD_IDP_DEVICE_HISTORY_RESPONSE_JSON);
+
+        // Compare the usage history from publisher A to that of publisher B
+        deviceAccessTestUtil.compareDeviceHistory(deviceHistoryFromPublisherA, deviceHistoryFromPublisherB);
+
+        // Remove the SAML entity from the device from publisher A
+        identityProviderTestUtil.removeIdpForDevice(publisherALocalId, publisherAToken, samlId);
+
+        // Get the usage history as publisher A and then publisher B
+        deviceHistoryFromPublisherA = deviceAccessTestUtil.testDeviceHistory(publisherALocalId, publisherAToken, AFTER_DELETE_IDP_DEVICE_HISTORY_RESPONSE_JSON);
+        deviceHistoryFromPublisherB = deviceAccessTestUtil.testDeviceHistory(publisherBLocalId, publisherBToken, AFTER_DELETE_IDP_DEVICE_HISTORY_RESPONSE_JSON);
+
+        // Ensure the usage history is the same for both publishers
+        deviceAccessTestUtil.compareDeviceHistory(deviceHistoryFromPublisherA, deviceHistoryFromPublisherB);
+
+        // Add back the SAML identity to the device
+        identityProviderTestUtil.testAddIdpToDeviceAndIdpResolution(5, publisherALocalId, publisherAToken, CREATE_SAML_IDP_REQUEST_JSON, CREATE_SAML_IDP_RESPONSE_JSON);
+
+        // Get the usage history as publisher A and then publisher B
+        deviceHistoryFromPublisherA = deviceAccessTestUtil.testDeviceHistory(publisherALocalId, publisherAToken, RE_ADD_SAML_IDP_HISTORY_RESPONSE_JSON);
+        deviceHistoryFromPublisherB = deviceAccessTestUtil.testDeviceHistory(publisherBLocalId, publisherBToken, RE_ADD_SAML_IDP_HISTORY_RESPONSE_JSON);
+
+        // Ensure the usage history is the same for both publishers
+        deviceAccessTestUtil.compareDeviceHistory(deviceHistoryFromPublisherA, deviceHistoryFromPublisherB);
+    }
+
+    @Test
+    public void testExistingDeviceDeletePublisherLocalId() {
```

Comment:
@balaf If you're curious, from here to the end of the file are the exact details of what I am testing